src/jestUtils.ts

```diff
@@ -0,0 +1,388 @@
+) => {
+  sendStateChangeEvent(
+    handler,
+    { oldState: 0, state: 2 },
+    baseEventData,
+    eventData
+  );
+};
+
+const sendProgressEventFirst: SendEventWrapper = (
+  handler,
+  baseEventData,
+  eventData
+) => {
+  sendStateChangeEvent(
+    handler,
+    { oldState: 2, state: 4 },
+    baseEventData,
+    eventData
+  );
+};
+
+const sendProgressEvent: SendEventWrapper = (
+  handler,
+  baseEventData,
+  eventData
+) => {
+  sendEvent(handler, { state: 4 }, baseEventData, eventData);
+};
+
+const sendEndEvent: SendEventWrapper = (handler, baseEventData, eventData) => {
+  sendStateChangeEvent(
+    handler,
+    { oldState: 4, state: 5 },
+    baseEventData,
+    eventData
+  );
+};
+
+const sendFailEvent: SendEventWrapper = (handler, baseEventData, eventData) => {
+  sendStateChangeEvent(
+    handler,
+    { oldState: 2, state: 1 },
+    baseEventData,
+    eventData
+  );
+};
+
+const sendCancelEvent: SendEventWrapper = (
+  handler,
+  baseEventData,
+  eventData
+) => {
+  sendStateChangeEvent(
+    handler,
+    { oldState: 4, state: 3 },
+    baseEventData,
+    eventData
+  );
+};
+
+const sendFinishEvent = (
+  handler: HandlerProperties,
+  baseEventData: Record<string, any>,
+  eventData?: FireGestureHandlerConfig<any>
+) => {
+  let states = { oldState: -1, state: -1 };
+  if (eventData?.onFail) {
+    states = { oldState: 2, state: 1 };
+  } else if (eventData?.onCancel) {
+    states = { oldState: 2, state: 2 };
+  } else {
+    states = { oldState: 2, state: 1 };
+  }
+  sendStateChangeEvent(handler, states, baseEventData, eventData?.onFinish);
+};
+
+const runEventsSequence = <T>(
+  component: any,
+  eventType: GestureHandlerType,
+  baseEventData: Record<string, any>,
+  eventData?: FireGestureHandlerConfig<T>
+) => {
+  const handlers = getGestureHandlers(eventType, component);
+  handlers.forEach((handler) => {
+    sendBeginEvent(handler, baseEventData, eventData?.configBegin);
+  });
+
+  handlers.forEach((handler) => {
+    baseEventData.handlerTag = handler.tag;
+    if (Array.isArray(eventData?.configProgress)) {
+      let index = 0;
+      for (const item of eventData!.configProgress) {
+        if (index === 0) {
+          sendProgressEventFirst(
+            handler,
+            baseEventData,
+            eventData?.configProgress
+          );
+        } else {
+          sendProgressEvent(handler, baseEventData, item);
+        }
+        index++;
+      }
+    } else {
+      sendProgressEventFirst(handler, baseEventData, eventData?.configProgress);
+    }
+  });
+
+  handlers.forEach((handler) => {
+    sendEndEvent(handler, baseEventData, eventData?.configEnd);
+  });
+
+  handlers.forEach((handler) => {
+    sendFailEvent(handler, baseEventData, eventData?.onFail);
+  });
+
+  handlers.forEach((handler) => {
+    sendCancelEvent(handler, baseEventData, eventData?.onCancel);
+  });
+
+  handlers.forEach((handler) => {
+    sendFinishEvent(handler, baseEventData, eventData);
+  });
```

Comment:
There is no need to handle `onFinish` separately. `useAnimatedGestureHandler` implements a state machine that calls it when it receives a `StateChangeEvent` where `oldState` is either `ACTIVE` or `BEGAN`, and the `state` is not (so it must be `END`, `FAILED`, `CANCELLED` or `UNDETERMINED`).